### ISSUE TITLE
[Feature] Support a custom logit processor

### ISSUE THREAD

**Issue by merrymercy** [sgl-project/sglang#2291]
We would like to support custom logit processors registered by users, so users can easily implement their custom sampling methods without the need to change the sglang code.

## Steps
1. After the following line, allowing running some user code to process the logits
https://github.com/sgl-project/sglang/blob/5f12f0e7af585973b366c262a5c3faea5ce5bf0a/python/sglang/srt/layers/sampler.py#L37
2. Add a new field `custom_params` to `SamplingParams`  https://github.com/sgl-project/sglang/blob/9a00e6f453e764c0b286e2a62f652a1202c0bf9c/python/sglang/srt/sampling/sampling_params.py#L43 so users can pass in their arguments. 

See also https://github.com/huggingface/transformers/pull/14779

**Comment by zhyncs** [comment]
BTW, constrained decoding and function calling can also be implemented this way, such as by crafting a mask kernel.
The schema part can continue to use outlines or xgrammar.

**Comment by zhyncs** [comment]
ref https://www.youtube.com/watch?v=I-a8sTMUq5o&t=3111s

**Comment by zhyncs** [comment]
cc @Ubospica

**Comment by gobraves** [comment]
Is anyone currently working on this issue? If not, can I give it a try?

**Comment by hongpeng-guo** [comment]
May I take this issue if no one is looking at it rn?

**Comment by merrymercy** [comment]
Supported by [#2396](https://github.com/sgl-project/sglang/pull/2396)
